### PR TITLE
Adding support for coarse grained gradual typing

### PR DESCRIPTION
**Pull request by deyaaeldeen**

### CODE REVIEW

File: Dynamizer/Lattice.hs [hunk 1]
```diff
@@ -64,15 +81,57 @@ class Gradual p where
...
 
+  funLattice (DConst name t e) = DConst name t <$> funLattice e
+  funLattice e@(DLam name args body t) =
+    DL.cons e $ DL.singleton $ DLam name args (dynamize body) $ dynamize t
+  funLattice e@(Lam args body t) = [e, Lam args (dynamize body) $ dynamize t]
+  funLattice (Bind name t e) = Bind name t <$> funLattice e
+  funLattice (As e t) = As <$> funLattice e <*> pure t
+  funLattice (Repeat x acc e1 e2 e3 e4 t) =
+    Repeat x acc <$> funLattice e1 <*> funLattice e2 <*> funLattice e3 <*> funLattice e4 <*> pure t
+  funLattice (Op op es) = Op op <$> mapM funLattice es
```
Comment (sabauma) [inline Dynamizer/Lattice.hs]: Maybe prefer mapA over mapM. Applicative appears sufficient for this example.

File: Dynamizer/Lattice.hs [hunk 2]
```diff
@@ -64,15 +81,57 @@ class Gradual p where
...
 
-instance (Gradual t, Gradual e) => Gradual (ExpF t e) where
+instance (Gradual t, Gradual e, Dynamize t, Dynamize e) => Gradual (ExpF t e) where
   lattice = bitraverse lattice pure
   count   = bifoldMap count count
   static  = bifoldMap static static
 
+  funLattice (DConst name t e) = DConst name t <$> funLattice e
+  funLattice e@(DLam name args body t) =
+    DL.cons e $ DL.singleton $ DLam name args (dynamize body) $ dynamize t
```
Comment (sabauma) [inline Dynamizer/Lattice.hs]: With the exception of the DLam case, all the other cases appear to to be equivalent to a traversable instance (might be the case we discussed). When I have dealt with cases like this, I would implement the special cases, then default to using traverse for everything else.

    funLattice e(@DLam name args body t) = ...
    funLattice e = traverse funLattice e

File: Dynamizer/Main.hs [hunk 3]
```diff
@@ -33,7 +33,7 @@ writeLattice b dname dps =
             return (n+1)) (0 :: Int) dps
 
 greet :: Options -> IO ()
-greet (Options srcFilePath ns nb) = do
+greet (Options srcFilePath ns nb coarce) = do
```
Comment (sabauma) [inline Dynamizer/Main.hs]: Should coarce be coerce?

File: Dynamizer/Lattice.hs (hunk 2 above)
Comment (deyaaeldeen) [reply Dynamizer/Lattice.hs]: done

File: Dynamizer/Main.hs (hunk 3 above)
Comment (deyaaeldeen) [reply Dynamizer/Main.hs]: should be coarse.

**Pull request merged**